### PR TITLE
Smart card redirection

### PR DESCRIPTION
**Pull request by eduardobeloni**
We think it's time to meld smart card redirection with the FreeRDP

### CODE REVIEW

File: channels/rdpdr/rdpdr_main.c [hunk 1]
```diff
@@ -573,7 +587,6 @@ rdpdr_process_irp(rdpdrPlugin * plugin, char* data, int data_size)
 
 	if (irp_get_event(&irp, &result) && irp.rwBlocking)
 	{
-		LLOGLN(10, ("IRP process pending events"));
```
Comment (otavio) [inline channels/rdpdr/rdpdr_main.c]: It seem unrelated. I think it is better to revert this change.

File: channels/rdpdr/rdpdr_main.h [hunk 2]
```diff
@@ -63,4 +65,5 @@ struct rdpdr_plugin
 	uint32 select_timeout;
 };
 
+
```
Comment (otavio) [inline channels/rdpdr/rdpdr_main.h]: Unrelated. Please drop it.

**Comment by otavio** [comment]
Besides the two above minor issues it seems ready for merging from my POV.

File: channels/rdpdr/smartcard/scard_operations.c [hunk 3]
```diff
@@ -0,0 +1,2030 @@
...
+			io->iffset += SERVER_READERSTATE_SIZE;
+		}
+
+		cur = readerStates;
+		curState = stateArray;
+		for (i = 0; i < readerCount; i++, cur++, curState++)
+		{
+			uint32 dataLength;
+
+#ifdef B_ENDIAN
```
Comment (otavio) [inline channels/rdpdr/smartcard/scard_operations.c]: The swap32 macro already handles the B_ENDIAN check so it seems duplication for me.

Another thing would be to rename the macro to something as little_endian_read_16 and little_endian_read_32 so it is clear that it handles the conversion in case a big endian machine is found.

File: channels/rdpdr/rdpdr_scard.c [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+	rdpdrPlugin * plugin = (rdpdrPlugin *)arg;
+
+	srv = devman_get_service_by_type(plugin->devman, RDPDR_DTYP_SMARTCARD);
+	if (!srv)
+	{
+		printf("%s: smart card service is not registered", __PRETTY_FUNCTION__);
+		return NULL;
+	}
+
+	while (1)
```
Comment (llyzs) [inline channels/rdpdr/rdpdr_scard.c]: In rdpdr_scard_finished_scanner, the thread is in an infinite loop which I think is going to be an issue. The thread must exit when the plugin is unloaded.

Comment (eduardobeloni) [inline channels/rdpdr/rdpdr_scard.c]: Can you take a look to the last commit?
I think this is enough to solve the issue.

**Comment by otavio** [comment]
@absmall we would like to squash your commits into the smart plugin commit. Does it sounds Ok for you?

We want to make it a single commit so it is easier to revert and like, if needed.

**Comment by absmall** [comment]
Squashing to one commit is fine with me

File: channels/rdpdr/rdpdr_main.c [hunk 5]
```diff
@@ -658,6 +672,7 @@ thread_process_message(rdpdrPlugin * plugin, char * data, int data_size)
 	uint16 packetID;
 	uint32 deviceID;
 	uint32 status;
+	SERVICE * srv;
```
Comment (otavio) [inline channels/rdpdr/rdpdr_main.c]: Please change srv to scard_srv so it is clear what service it is handling.

File: channels/rdpdr/rdpdr_scard.h [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+#ifndef __RDPDR_SCARD_H
+#define __RDPDR_SCARD_H
+
+#include <pthread.h>
+
+#include "rdpdr_types.h"
+#include "rdpdr_main.h"
+
+
```
Comment (otavio) [inline channels/rdpdr/rdpdr_scard.h]: Drop one of new lines.

File: channels/rdpdr/smartcard/scard_main.c [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+*/
+
+#include <stdlib.h>
+
+#include "rdpdr_types.h"
+#include "rdpdr_constants.h"
+
+#include "scard_main.h"
+
+
```
Comment (otavio) [inline channels/rdpdr/smartcard/scard_main.c]: Same here, please drop a line.

File: channels/rdpdr/smartcard/scard_operations.c [hunk 8]
```diff
@@ -0,0 +1,2069 @@
...
+#ifdef WITH_DEBUG_SCARD
+#include <libfreerdp/frdp.h>
+#endif
+
+#include "rdpdr_types.h"
+#include "rdpdr_constants.h"
+
+#include "scard_main.h"
+
+
```
Comment (otavio) [inline channels/rdpdr/smartcard/scard_operations.c]: Same here.

File: channels/rdpdr/smartcard/scard_operations.c [hunk 9]
```diff
@@ -0,0 +1,2069 @@
...
+#define SCARD_IOCTL_TRANSMIT                 0x000900D0	/* Transmit */
+#define SCARD_IOCTL_CONTROL                  0x000900D4	/* Control */
+#define SCARD_IOCTL_GETATTRIB                0x000900D8	/* GetAttrib */
+#define SCARD_IOCTL_SETATTRIB                0x000900DC	/* SetAttrib */
+#define SCARD_IOCTL_ACCESS_STARTED_EVENT     0x000900E0	/* SCardAccessStartedEvent */
+#define SCARD_IOCTL_LOCATE_CARDS_BY_ATR      0x000900E8	/* LocateCardsByATR */
+
+#define SCARD_INPUT_LINKED                   0xFFFFFFFF
+
+
```
Comment (otavio) [inline channels/rdpdr/smartcard/scard_operations.c]: Same here.

File: channels/rdpdr/smartcard/scard_operations.c [hunk 10]
```diff
@@ -0,0 +1,2069 @@
...
+static ScardQueue *pending_queue = NULL;
+static pthread_mutex_t pending_guard;
+static pthread_cond_t pending_ready;
+
+static ScardQueue *finished_queue = NULL;
+static pthread_mutex_t finished_guard;
+static sem_t finished_ready;
+
+
```
Comment (otavio) [inline channels/rdpdr/smartcard/scard_operations.c]: And here.

File: channels/rdpdr/smartcard/scard_operations.c [hunk 11]
```diff
@@ -0,0 +1,2069 @@
...
+static DWORD
+handle_Status(struct io_wrapper *io, RD_BOOL wide);
+
+static LONG
+handle_State(struct io_wrapper *io);
+
+static LONG
+handle_GetAttrib(struct io_wrapper *io);
+
+
```
Comment (otavio) [inline channels/rdpdr/smartcard/scard_operations.c]: Here too.

File: channels/rdpdr/smartcard/scard_operations.c [hunk 12]
```diff
@@ -0,0 +1,2069 @@
...
+			default:
+				sc_handle_async_request(irp);
+				break;
+		}
+	}
+	return NULL;
+}
+
+
```
Comment (otavio) [inline channels/rdpdr/smartcard/scard_operations.c]: Extra line here too.

File: channels/rdpdr/smartcard/scard_operations.c [hunk 13]
```diff
@@ -0,0 +1,2069 @@
...
+	int header = (length < 0) ? (0) : ((length > highLimit) ? (highLimit) : (length));
+
+	DEBUG_SCARD("\n");
+
+	SET_UINT32(io->outbuf, io->offset, header);
+	SET_UINT32(io->outbuf, io->offset + 4, 0x00000001);	/* Magic DWORD - any non zero */
+	io->offset += 8;
+}
+
+
```
Comment (otavio) [inline channels/rdpdr/smartcard/scard_operations.c]: Guess what? :-)

**Comment by otavio** [comment]
Hello folks ... Eduardo has fixed the minor issues we have found here and also done the changes suggested by @llyzs so we believe this can be merged into master.

Do someone objects to it?

**Comment by otavio** [comment]
I have fixed few warnings on smartcard code and pushed it.

Let's see how it behave :-)
